File: Classes/BTInterpolator.c
//
// Betwixt - Copyright 2012 Three Rings Design

#include "BTInterpolator.h"

#define MIN(a,b) (((a)<(b))?(a):(b))

BTInterpolator BTLinearInterpolator = ^ float (float ratio) { return ratio; };
BTInterpolator BTEaseInInterpolator = ^ float (float ratio) { return ratio * ratio * ratio; };
BTInterpolator BTEaseOutInterpolator = ^ float (float ratio) {
    float inverse = ratio - 1.0f;
    return inverse * inverse * inverse + 1.0f;
};
BTInterpolator BTEaseInOutInterpolator = ^ float (float ratio) {
    if (ratio < 0.5f) return 0.5f * BTEaseOutInterpolator(ratio * 2.0f);
    else  return 0.5f * BTEaseInInterpolator((ratio - 0.5f) * 2.0f) + 0.5f;
};

float BTInterpolate (float from, float to, float elapsedTime, float totalTime, 
                     BTInterpolator interp) {
    elapsedTime = MIN(elapsedTime, totalTime);
    return from + ((to - from) * interp(elapsedTime / totalTime));
}

File: Classes/BTInterpolator.m
//
// Betwixt - Copyright 2012 Three Rings Design

#import "BTInterpolator.h"

@interface BTLinearInterpolator : BTInterpolator
@end

@implementation BTLinearInterpolator
- (float)interpolate:(float)from to:(float)to dt:(float)dt t:(float)t {
    if (t == 0) { 
        return to;
    }
    return from + ((to - from) * (dt / t));
}
@end

@interface BTEaseInInterpolator : BTInterpolator
@end

@implementation BTEaseInInterpolator
- (float)interpolate:(float)from to:(float)to dt:(float)dt t:(float)t {
    if (t == 0) { 
        return to;
    }
    float dtt = dt / t;
    return from + ((to - from) * dtt * dtt * dtt);
}
@end

@interface BTEaseOutInterpolator : BTInterpolator
@end

@implementation BTEaseOutInterpolator
- (float)interpolate:(float)from to:(float)to dt:(float)dt t:(float)t {
    if (t == 0) { 
        return to;
    }
    float dtt = dt / t - 1;
    return from + ((to - from) * (1 + dtt * dtt * dtt));
}
@end

@interface BTEaseInOutInterpolator : BTInterpolator
@end

@implementation BTEaseInOutInterpolator
- (float)interpolate:(float)from to:(float)to dt:(float)dt t:(float)t {
    if (t == 0) { 
        return to;
    }
    float range = to - from;
    float hdtt = dt / (t/2);
    if (hdtt < 1) {
        return from + range/2 * hdtt * hdtt * hdtt;
    } else {
        float nhdtt = hdtt - 2;
        return from + range/2 * (2 + nhdtt * nhdtt * nhdtt);
    }
}
@end

@implementation BTInterpolator
- (float)interpolate:(float)from to:(float)to dt:(float)dt t:(float)t { return to; }

+ (BTInterpolator*)LINEAR {
    static BTInterpolator* interp = nil;
    @synchronized(self) {
        if (interp == nil) {
            interp = [[BTLinearInterpolator alloc] init];
        }
    }
    return interp;
}

+ (BTInterpolator*)EASE_IN {
    static BTInterpolator* interp = nil;
    @synchronized(self) {
        if (interp == nil) {
            interp = [[BTEaseInInterpolator alloc] init];
        }
    }
    return interp;
}

+ (BTInterpolator*)EASE_OUT {
    static BTInterpolator* interp = nil;
    @synchronized(self) {
        if (interp == nil) {
            interp = [[BTEaseOutInterpolator alloc] init];
        }
    }
    return interp;
}

+ (BTInterpolator*)EASE_IN_OUT {
    static BTInterpolator* interp = nil;
    @synchronized(self) {
        if (interp == nil) {
            interp = [[BTEaseInOutInterpolator alloc] init];
        }
    }
    return nil;
}
@end